src/compute-client/src/metrics.rs

```diff
@@ -0,0 +1,126 @@
+
+//! Metrics for the compute controller components
+
+use std::sync::Arc;
+
+use mz_ore::cast::{CastFrom, TryCastFrom};
+use mz_ore::metric;
+use mz_ore::metrics::DeleteOnDropHistogram;
+use mz_ore::metrics::HistogramVecExt;
+use mz_ore::metrics::MetricsRegistry;
+use mz_service::codec::StatsCollector;
+
+use crate::controller::ComputeInstanceId;
+use crate::controller::ReplicaId;
+use crate::protocol::command::ProtoComputeCommand;
+use crate::protocol::response::ProtoComputeResponse;
+
+/// Compute controller metrics
+#[derive(Debug, Clone)]
+pub struct ComputeControllerMetrics {
+    messages_sent_bytes: prometheus::HistogramVec,
+    messages_received_bytes: prometheus::HistogramVec,
+}
+
+impl ComputeControllerMetrics {
+    pub fn new(metrics_registry: MetricsRegistry) -> Self {
+        ComputeControllerMetrics {
+            messages_sent_bytes: metrics_registry.register(metric!(
+                name: "mz_compute_messages_sent_bytes",
+                help: "size of compute messages sent",
+                var_labels: ["instance", "replica"]
+            )),
+            messages_received_bytes: metrics_registry.register(metric!(
+                name: "mz_compute_messages_received_bytes",
+                help: "size of compute messages received",
+                var_labels: ["instance", "replica"]
+            )),
+        }
+    }
+
+    pub fn for_instance(&self, instance_id: ComputeInstanceId) -> InstanceMetrics {
+        InstanceMetrics {
+            messages_sent_bytes: self.messages_sent_bytes.clone(),
+            messages_received_bytes: self.messages_received_bytes.clone(),
+            instance_id,
+        }
+    }
+}
+
+/// Per-instance metrics
+#[derive(Debug, Clone)]
+pub struct InstanceMetrics {
+    messages_sent_bytes: prometheus::HistogramVec,
+    messages_received_bytes: prometheus::HistogramVec,
+    instance_id: ComputeInstanceId,
+}
+
+impl InstanceMetrics {
+    pub fn for_replica(&self, replica_id: ReplicaId) -> ReplicaMetrics {
+        let labels = vec![self.instance_id.to_string(), replica_id.to_string()];
+        let messages_sent_bytes = self
+            .messages_sent_bytes
+            .get_delete_on_drop_histogram(labels.clone());
+        let messages_received_bytes = self
+            .messages_received_bytes
+            .get_delete_on_drop_histogram(labels);
+
+        ReplicaMetrics {
+            inner: Arc::new(ReplicaMetricsInner {
+                messages_sent_bytes,
+                messages_received_bytes,
+            }),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ReplicaMetricsInner {
+    messages_sent_bytes: DeleteOnDropHistogram<'static, Vec<String>>,
+    messages_received_bytes: DeleteOnDropHistogram<'static, Vec<String>>,
+}
+
+/// Per-replica metrics.
+#[derive(Debug, Clone)]
+pub struct ReplicaMetrics {
+    inner: Arc<ReplicaMetricsInner>,
+}
+
+impl ReplicaMetrics {
+    pub fn for_connection(&self, _index: usize) -> ReplicaConnectionMetrics {
```

Comment:
What's the purpose of the unused argument?

Reply:
If we have a multi process replica, we could differentiate on the process. But we currently agreed on collecting the metrics per-replica (see [slack discussion](https://materializeinc.slack.com/archives/CM7ATT65S/p1673884428820379)).